### ISSUE TITLE
Relation between AVX and TH?

### ISSUE THREAD

**Issue by jiseongg** [pytorch/pytorch#25171]
Hi, I'm studying PyTorch internals, especially trying to find AVX implementation. (not AVX2.)

I built PyTorch from source with
```bash
DEBUG=1 NO_CUDA=1 python setup.py develop
```
What I've figured out is the AVX intrinsic function that is called for simple tensor addition.

```python
z = torch.add(x, y)
```
I debugged with gdb, and set some breakpoints and succeeded it. Below is the result of `bt` at gdb.

```
#0 ~/pytorch/aten/src/ATen/cpu/vec256/vec256_double.h:203
...
#18 ~/pytorch/build/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.AVX.cpp:29
...
#20 ~/pytorch/aten/src/ATen/native/BinaryOps.cpp:41
```
The AVX intrinsic `_mm256_add_pd(a, b)` is called at #0

Now, I'm trying to find if there exists AVX acceleration of CNN, so I've debugged really simple code at [PyTorch tutorial](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html#sphx-glr-beginner-blitz-cifar10-tutorial-py) with same procedure as above. But I cannot find any AVX implementation.

So now I'm wondering if I've missed some options when I built PyTorch. I've googled a lot, and looking inside the CMake system, I haven't figured it out yet.
One I noticed it is there are AVX.cpp in TH directory.
`~/pytorch/aten/src/TH/vector/AVX.cpp`
And after searching TH, I also noticed that I don't have libTH.so file. In [a blog in PyTorch homepage](https://pytorch.org/blog/a-tour-of-pytorch-internals-2/#backend-torch-and-vendor-libraries), there is libTH.so.1 at torch/lib. Below is my `ls` result at that directory.
```bash
user@user:~/pytorch/torch/lib
c10d         libcaffe2_detectron_ops.so        libfbgemm.a     libprotobuf-lited.a  libshm.so           pkgconfig
cmake        libcaffe2_module_test_dynamic.so  libgloo.a       libprotocd.a         libshm_windows      python3.6
libasmjit.a  libcaffe2_observers.so            libmkldnn.a     libpthreadpool.a     libsleef.a
libc10d.a    libclog.a                         libnnpack.a     libqnnpack.a         libtorch_python.so
libc10.so    libcpuinfo.a                      libprotobufd.a  libshm               libtorch.so
```

In addition, in [this blog](https://apaszke.github.io/torch-internals.html), there is "simd" directory in `TH/generic`, but I don't have it. Is there anyone who can tell me about what I'm missing?

**Comment by jiseongg** [comment]
@mrshenli, 
> I did some search, and found that convolutions on CPU calls `thnn_conv2d` and `thnn_conv3d`.
> ...
> Which seems calls into `THTensorConv.cpp`, which does use the vectorized implementation.

**Comment by jiseongg** [comment]
@mrshenli,
 
> I did some search, and found that convolutions on CPU calls `thnn_conv2d` and `thnn_conv3d`.
> ...
> Which seems calls into `THTensorConv.cpp`, which does use the vectorized implementation.

I'm not sure that tutorial code calls it. Functions in `THTensorConv.cpp` don't exist at all in my call stack. But I will do more research about it.

>I believe that's all merged into libtorch.so now. If you do nm libtorch.so you should be able to see THNN_ symbols.

It seems you're right. I found symbols of source files at `TH` directory. Thank you. But I think my codes don't reach there, as I mentioned above.


@colesbury,

Thank you, I set a breakpoint at that function and it succeeded. I've never known CPU executes PyTorch with mkl-dnn.


@vpirogov,

Yes, It seems to be what I was looking for. But I'm not sure that my code execute it, so I'm trying to go deeper with my python code to figure out if it reach there.
